### PR TITLE
Fast forwarding 2.2 branch to latest master branch

### PR DESCRIPTION
**Pull request by kumare3**
There are few fixes in master that should be pulled into 2.2
Also, this will set the stage to make the larger change on top of this branch.

### CODE REVIEW

**Comment by kumare3** [comment]
I signed it!

**Comment by kumare3** [comment]
@liyinan926 can you please approve the merge? This is just fast forwarding to the latest master. We have tested this, and barring the race condition bug, which we will fix through the refactor it works just as expected

File: Dockerfile [hunk 1]
```diff
@@ -14,6 +14,24 @@
...
+RUN chmod +x /usr/bin/dep
+
+WORKDIR ${GOPATH}/src/k8s.io/spark-on-k8s-operator
+COPY Gopkg.toml Gopkg.lock ./
+RUN dep ensure -vendor-only
+COPY . ./
+RUN go generate && go build -o /usr/bin/spark-operator
+
+
+FROM gcr.io/ynli-k8s/spark:v2.3.0
```
Comment (liyinan926) [inline Dockerfile]: This should be `FROM kubespark/spark-base:v2.2.0-kubernetes-0.5.0` for 2.2.

File: Dockerfile [hunk 2]
```diff
@@ -14,6 +14,24 @@
...
+COPY --from=builder /usr/bin/spark-operator /usr/bin/
+>>>>>>> master
```
Comment (liyinan926) [inline Dockerfile]: Remove this.

File: Dockerfile [hunk 3]
```diff
@@ -14,6 +14,24 @@
 # limitations under the License.
 #
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (kumare3) [reply Dockerfile]: hey good catch, will do

[the remark this answers is not in the document]
Comment (kumare3) [reply Dockerfile]: Also I will follow up with a commit to use scratch to build the go binary. That makes the image really tiny

[the remark this answers is not in the document]
Comment (kumare3) [reply Dockerfile]: updated

File: Dockerfile (hunk 2 above)
Comment (kumare3) [reply Dockerfile]: done

File: Dockerfile (hunk 1 above)
Comment (kumare3) [reply Dockerfile]: done

**Comment by liyinan926** [comment]
@kumare3 Looks like this is still missing a few commits made to the master branch.

**Comment by kumare3** [comment]
@liyinan926 thank you for bringing that to my attention. I have pulled the recent tip. Can you check and approve?

**Pull request merged**